nni/compression/pytorch/speedup/v2/dependency.py

```diff
@@ -0,0 +1,444 @@
+]
+
+# see https://pytorch.org/docs/stable/torch.html#pointwise-ops
+CALL_FUNCTION_REDUCE = [
+    torch.add, torch.sub, torch.subtract, torch.mul, torch.div, torch.multiply, torch.divide,
+    torch.addcmul, torch.addcdiv, torch.logical_xor, torch.logical_and, torch.logical_or,
+    operator.add, operator.iadd, operator.sub, operator.isub, operator.mul, operator.imul,
+    operator.truediv, operator.floordiv, operator.or_, operator.and_, operator.xor,
+]
+CALL_METHOD_REDUCE = [
+    'add', 'add_', 'sub', 'sub_', 'subtract', 'subtract_', 'mul', 'mul_',
+    'div', 'div_', 'multiply', 'multiply_', 'divide', 'divide_',
+    'addcmul', 'addcmul_', 'addcdiv', 'addcdiv_', 'logical_xor', 'logical_xor_',
+    'logical_and', 'logical_and_', 'logical_or', 'logical_or_',
+]
+
+# see https://pytorch.org/docs/stable/torch.html#indexing-slicing-joining-mutating-ops
+CALL_FUNCTION_CONCAT = [torch.cat, torch.concat, torch.concatenate, torch.column_stack, torch.dstack,
+                     torch.hstack, torch.row_stack, torch.stack]
+CALL_FUNCTION_RESHAPE = [
+    *CALL_FUNCTION_CONCAT,
+    torch.chunk, torch.diagonal, torch.flatten, torch.flip, torch.fliplr, torch.flipud, torch.moveaxis, torch.movedim, torch.narrow,
+    torch.reshape, torch.split, torch.split_with_sizes, torch.squeeze, torch.unsqueeze, torch.transpose, torch.t, torch.permute,
+    torch.repeat_interleave,
+]
+CALL_METHOD_RESHAPE = [
+    'chunk', 'diagonal', 'flatten', 'flip', 'fliplr', 'flipud', 'moveaxis', 'movedim', 'narrow',
+    'reshape', 'split', 'split_with_sizes', 'squeeze', 'unsqueeze', 'transpose', 't', 'permute',
+    'view', 'view_as', 'expand', 'expand_as', 'expand_dims', 'repeat', 'repeat_interleave',
+]
+logger = logging.getLogger('Shape Dependency')
+
+
+def lcm_list(L):
+    lcm = 1
+    for i in L:
+        lcm = np.lcm(lcm, i)
+    return lcm
+
+
+def gcd_list(L):
+    gcd = L[0]
+    for i in L:
+        gcd = np.gcd(gcd, i)
+    return gcd
+
+
+def channel_dependency_breakpoint(node: torch.fx.Node):
+    """
+    Check if this operation will break the channel dependency.
+
+    Parameters
+    ----------
+    node : torch.fx.Node
+
+    Returns
+    -------
+    bool
+        If this operation will break the channel dependency.
+    """
+
+    def find_parent(node: torch.fx.Node):
+        par_arg = node.all_input_nodes[0]
+
+        while isinstance(par_arg, tuple):
+            par_arg = par_arg[0]
+        return par_arg
+
+    def get_shape(node):
+        meta = node.meta.get('tensor_meta', None)
+        if isinstance(meta, TensorMetadata):
+            return meta[0]
+        elif isinstance(meta, tuple):
+            return meta[0][0]
+        return None
+
+    in_shape = get_shape(find_parent(node))
+    out_shape = get_shape(node)
+    if not in_shape or not out_shape or len(in_shape) <= 1 or len(out_shape) <= 1:
+        return True
+    in_channel = in_shape[1]
+    out_channel = out_shape[1]
+    return in_channel != out_channel
```

Comment:
Is it possible that the number of channels is the same but it is a break operation?

Reply:
it's really hard for me to find an example. maybe I add a TODO here?